spec/abci++/abci++_client_server_002_draft.md

```diff
@@ -0,0 +1,110 @@
+This means ABCI applications written in Golang can be compiled with Tendermint Core and run as a single binary.
+
+### GRPC
+
+If GRPC is available in your language, this is the easiest approach,
+though it will have significant performance overhead.
+
+To get started with GRPC, copy in the [protobuf
+file](../../proto/tendermint/abci/types.proto) and compile it using the GRPC
+plugin for your language. For instance, for Golang, the command is `protoc
+--go_out=plugins=grpc:. types.proto`.  See the [grpc documentation](http://www.grpc.io/docs/)
+for more details.  `protoc` will autogenerate all the
+necessary code for ABCI client and server in your language, including whatever
+interface your application must satisfy to be used by the ABCI server for
+handling requests.
+
+Note the length-prefixing used in the socket implementation (TSP) does not apply for GRPC.
+
+### TSP
+
+Tendermint Socket Protocol is an asynchronous, raw socket server which provides ordered message passing over unix or tcp.
+Messages are serialized using Protobuf3 and length-prefixed with a [signed Varint](https://developers.google.com/protocol-buffers/docs/encoding?csw=1#signed-integers)
+
+If GRPC is not available in your language, or you require higher
+performance, or otherwise enjoy programming, you may implement your own
+ABCI server using the Tendermint Socket Protocol. The first step is still to auto-generate the relevant data
+types and codec in your language using `protoc`. In addition to being proto3 encoded, messages coming over
+the socket are length-prefixed to facilitate use as a streaming protocol. proto3 doesn't have an
+official length-prefix standard, so we use our own. The first byte in
+the prefix represents the length of the Big Endian encoded length. The
+remaining bytes in the prefix are the Big Endian encoded length.
+
+For example, if the proto3 encoded ABCI message is 0xDEADBEEF (4
+bytes), the length-prefixed message is 0x0104DEADBEEF. If the proto3
+encoded ABCI message is 65535 bytes long, the length-prefixed message
+would start with 0x02FFFF.
+
+The benefit of using this `varint` encoding over the old version (where integers were encoded as `<len of len><big endian len>` is that
+it is the standard way to encode integers in Protobuf. It is also generally shorter.
+
+As noted above, this prefixing does not apply for GRPC.
+
+An ABCI server must also be able to support multiple connections, as
+Tendermint uses four connections.
+
+## Client
+
+There are currently two use-cases for an ABCI client. One is a testing
+tool, as in the `abci-cli`, which allows ABCI requests to be sent via
+command line. The other is a consensus engine, such as Tendermint Core,
+which makes requests to the application every time a new transaction is
+received or a block is committed.
+
+It is unlikely that you will need to implement a client. For details of
+our client, see
+[here](https://github.com/tendermint/tendermint/tree/master/abci/client).
```

Comment:
Which client is this? The one used by `abci-cli`, by the Core of both?

Reply:
It points to the boilerplate we have to route ABCI request to the Application, depending the the server type (in-process, GRPC, socket).
I've removed this paragraph.